planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/MbtilesWriter.java

```diff
@@ -225,43 +233,131 @@ private void tileEncoderSink(Iterable<TileBatch> prev) throws IOException {
+  private interface TileFeaturesEncoder {
+    TileEncodingResult encode(FeatureGroup.TileFeatures tileFeatures) throws IOException;
+  }
+
+  private abstract class TileFeaturesEncoderBase implements TileFeaturesEncoder {
+    protected byte[] encodeVectorTiles(FeatureGroup.TileFeatures tileFeatures) {
+      VectorTile en = tileFeatures.getVectorTileEncoder();
+      byte[] encoded = en.encode();
+      if (encoded.length > 1_000_000) {
+        LOGGER.warn("{} {}kb uncompressed", tileFeatures.tileCoord(), encoded.length / 1024);
+      }
+      writeVectorTilesEncodingStats(tileFeatures, encoded);
+      return encoded;
+    }
+
+    protected void writeVectorTilesEncodingStats(FeatureGroup.TileFeatures tileFeatures, byte[] encoded) {
+      int zoom = tileFeatures.tileCoord().z();
+      int encodedLength = encoded == null ? 0 : encoded.length;
+      totalTileSizesByZoom[zoom].incBy(encodedLength);
+      maxTileSizesByZoom[zoom].accumulate(encodedLength);
+    }
+  }
+
+  private class DefaultTileFeaturesEncoder extends TileFeaturesEncoderBase {
+
     /*
      * To optimize emitting many identical consecutive tiles (like large ocean areas), memoize output to avoid
      * recomputing if the input hasn't changed.
      */
-    byte[] lastBytes = null, lastEncoded = null;
+    private byte[] lastBytes = null, lastEncoded = null, lastBytesForHasing = null;
+
+    @Override
+    public TileEncodingResult encode(FeatureGroup.TileFeatures tileFeatures) throws IOException {
+
+      byte[] bytes, encoded, currentBytesForHashing;
+
+      currentBytesForHashing = tileFeatures.getBytesRelevantForHashing();
+
+      if (Arrays.equals(lastBytesForHasing, currentBytesForHashing)) {
+        bytes = lastBytes;
+        encoded = lastEncoded;
+        writeVectorTilesEncodingStats(tileFeatures, encoded);
+        memoizedTiles.inc();
+      } else {
+        encoded = lastEncoded = encodeVectorTiles(tileFeatures);
+        bytes = lastBytes = gzip(encoded);
+        lastBytesForHasing = currentBytesForHashing;
+      }
+      return new TileEncodingResult(tileFeatures.tileCoord(), bytes, -1);
+    }
+  }
+
+  private class CompactTileFeaturesEncoder extends TileFeaturesEncoderBase {
+
+    /*
+     * to put some context on this number, duplicates in Australia per "NumFeaturesToEmit":
+     * 1=13074826, 2=1359105, 3=397821, 4=13822, 5=73, 7=4, 8=2, 9=2, 18=2
+     */
+    private static final int MAX_FEATURES_HASHING_THRESHOLD = 5;
+
+    private static final int FNV1_32_INIT = 0x811c9dc5;
+    private static final int FNV1_PRIME_32 = 16777619;
+
+    private final Supplier<Long> tileDataIdGenerator;
+    private final Map<Integer, Long> tileDataIdByHash = new HashMap<>((int) MAX_TILES_PER_BATCH);
+
+    private CompactTileFeaturesEncoder(Supplier<Long> tileDataIdGenerator) {
+      this.tileDataIdGenerator = tileDataIdGenerator;
+    }
+
+    @Override
+    public TileEncodingResult encode(TileFeatures tileFeatures) throws IOException {
```

Comment:
Can we move the tile cache to the single-threaded writer so that we deduplicate a smaller number of entries for the unique tiles?  The breakdown of work would roughly be:

`encode` (in workers)
- (only when using --compact-db) if num feature is < `MAX_FEATURES_HASHING_THRESHOLD` then compute the hash and pass it along on `TileEncodingResult` otherwise some value indicating "not repeated"
- use the old `hasSameContents ` logic to avoid encoding a tile if it's the same as the last tile - I think that should be more than sufficient to avoid unnecessary work since it's a cheap check and will likely avoid 99%+ of the repeat encoding work. We could also use the hasSameContents check to memoize the hash between identical tiles.

`tileWriter` / compact batch writer (single-threaded)
- keep a shared hppc `IntIntHashMap` (?) from tile hash to tile ID, and append new values to it when we encounter a repeated tile where the hash was computed in encode
- append to the `tiles_data` table except when you encounter a repeated tile where we've already added the hash to it.

Reply:
Oookay, so far I was under the assumption that the current tile memoization isn't that effective. Thus I tried to knock-off 2 things with the hashmap, de-duping for writing and avoiding to re-encode+gzip. 
I'll try to verify the effectiveness of the current memoization and if it's that effective, then I'll refactor accordingly.

Reply:
I think the current memoization is pretty effective since the tiles are processed in order in consecutive batches, but definitely let me know if you find otherwise. I tested Australia on my m1 mac without memoization and encode time went from 140 seconds to 370 cpu seconds.

Reply:
addressed in https://github.com/onthegomap/planetiler/pull/219/commits/5414f33ddd27d0889c0efbd8d85ca0b1b24e16ee

I left a comment in code here, https://github.com/onthegomap/planetiler/pull/219/commits/5414f33ddd27d0889c0efbd8d85ca0b1b24e16ee#diff-38e31206bb20aac6140acd652533b5194401ffbf1d95ede4cb620d2294bc40a8R673

I think by trying to de-dupe memoized only we miss quite some opportunities and I think we should just try to de-dupe whenever there's a hash - but I understand if you want to avoid potential unnecessary map lookups

results on my machine
```

memoized, only
[m[0m0:06:13 INF - 	mbtiles          1m9s cpu:10m46s gc:4s avg:9.4
[m[0m0:06:13 INF - 	  read    1x(22% 15s sys:2s wait:46s done:3s)
[m[0m0:06:13 INF - 	  encode 12x(63% 43s wait:13s done:3s)
[m[0m0:06:13 INF - 	  write   1x(56% 38s sys:4s wait:21s)
[m[0m0:06:13 INF - ----------------------------------------
[m[0m0:06:13 INF - 	features	3.6GB
[m[0m0:06:13 INF - 	mbtiles	1.8GB

1814077440

all that have a hash
[m[0m0:06:11 INF - 	mbtiles          1m7s cpu:10m28s gc:4s avg:9.3
[m[0m0:06:11 INF - 	  read    1x(22% 15s sys:2s wait:46s done:3s)
[m[0m0:06:11 INF - 	  encode 12x(62% 42s wait:13s done:3s)
[m[0m0:06:11 INF - 	  write   1x(54% 37s sys:3s wait:22s)
[m[0m0:06:11 INF - ----------------------------------------
[m[0m0:06:11 INF - 	features	3.6GB
[m[0m0:06:11 INF - 	mbtiles	1.7GB

1725030400
```

Reply:
sounds good! I think it's fine to always do the map lookup in the writer.